RFC: Stable API deprecation: SPI: Add timing and other parameters to spi peripherals specs from DT

decsny opened the pull request:
- Problem Description:
   1) SPI transaction timing (and some other) parameters mostly come from the SPI bus peripherals, not the controller. This information is found by looking at the datasheets for the SPI devices. These are hardware requirements and should be specified on the bus peripheral nodes in DT.
   2) Currently, delay value can only be provided for a GPIO CS in microseconds. This is, in almost all cases, not useful if the timing is really important, because almost all the spi device timing requirements for these parameters are on the order of a few nanoseconds. Many SPI controller have the ability to natively handle these timings without a GPIO CS, so this RFC is giving a way to configure the actual controller hardware properly. Especially for small transfers, if maximum resolution on these parameter is in 1 microsecond increments, but the transfer is only one byte which can be transferred in nanoseconds on the bus, then this is a huge waste of time on utilizing the SPI bus efficiently.
- Proposed Change:
   Add the parameters to the common spi-device.yaml DT binding file, based on same linux parameters. Get these values from DT in the SPI DT macros. Add to the spi config structs to be able to pass this info to the drivers through the SPI API calls. 
- Detailed RFC:
  - Devicetree additions to spi-device.yaml can be viewed here: https://github.com/zephyrproject-rtos/zephyr/blob/288bda74396aaccae33b8221541354f71aa85caf/dts/bindings/spi/spi-device.yaml . These properties come from linux: https://www.kernel.org/doc/Documentation/devicetree/bindings/spi/spi-peripheral-props.yaml . Some specific spi device DT bindings were also edited to put a default value for the delay parameters if the corresponding driver was hardcoding a delay value using the old macro.
  - All the other changes are not to function signatures but just structures. Usually this struct definition is transparent to user if they are using the DT macros, which is virtually all the cases in tree. And the changes are meant to be backwards compatible, the old structs are a union with the new one. So 

- Dependencies:
  Impact to users of the API will be that they just need to remove the third argument (delay) from the SPI DT macros, as this will now automatically get retrieved from DT by the same macros, so this argument is no longer necessary. If they don't make this change then there will be a build-time preprocessor error. The rest of the API change was intentionally designed to be backward compatible and not to cause any other breakages.


Expand the struct spi_cs_control to be able to specify hardware timing parameters for a SPI peripheral, primarily intended to  be generated from properties in DT on the spi device nodes, inspired by linux binding at https://www.kernel.org/doc/Documentation/devicetree/bindings/spi/spi-peripheral-props.yaml

Support these parameters in NXP LPSPI driver

If the PR approach is okay, then TODO on this PR:
- [x] Figure  out what new testing is needed
- [x] Documentation review
- [x] Update release notes / migration guide about changes to multiple things that are consumers of SPI in addition to SPI changes themselves.
- [x] Update driver API version?
- [x] Fixing any CI issues

tbursztyka commented on include/zephyr/drivers/spi.h:
```diff
@@ -152,24 +152,49 @@ extern "C" {
+ * to use GPIO or native hardware CS, and timing parameters.
  *
  */
 struct spi_cs_control {
```
Perhaps it's time to rename this to spi_cs_and_delay_control

decsny replied:
renaming would be an API change instead of an extension though, are you sure it's worth doing that, as opposed to just having a couple extra bytes in the overall spi_config struct (whether the way I have in this PR or introduce a new spi_timing struct)

tbursztyka replied:
If we need an API change (whatever that could be), so be it. Better having a sane API rather than a half-hacked one.

tbursztyka commented on include/zephyr/drivers/spi.h:
```diff
@@ -329,8 +354,10 @@ struct spi_config {
-};
 
+	/** @brief Delay between data frames in nanoseconds (0 means 50% of frequency) */
+	uint16_t dfs_delay_ns;
```
and this could go into the anonymous struct you added, in the union.

That way: to change in structure size.

That said, this is very much spi controller specific. Some controller do not offer anything to configure this. 
And 0 as 50% of the frequency, why? why not just "hardware default"?

decsny replied:
I thought about this too but since the struct was called spi_cs_control and this was not related I thought it would be confusing

tbursztyka commented on include/zephyr/drivers/spi.h:
```diff
@@ -152,24 +152,49 @@ extern "C" {
-	 * Delay in microseconds to wait before starting the
-	 * transmission and before releasing the CS line.
+	 * Timing configuration.
+	 * For GPIO CS, delay is microseconds to wait before starting
```
this was not only meant for gpio cs actually. thus  why the structure was called spi_cs_control, and not spi_gpio_cs_control (for instance)

decsny replied:
yeah, that was convenient that it was already named generically, I noticed

tbursztyka reviewed (changes requested):
This needs further thoughts.

For instance, you set the size of the ns delay to uint8_t, will that be sufficient in all cases?

Another issue as well: how device drivers are going to know which from the delay to the anonymous struct to use? there is no flag anywhere for mitigation. Unless we want to move to a new delay system, which case we will need to make delay attribute obsolete etc...

a few additional comments inline

decsny commented:
> This needs further thoughts.
> 
> For instance, you set the size of the ns delay to uint8_t, will that be sufficient in all cases?

I think almost all data sheets I have seen have CS delay parameters of < 10 ns, but if someone knows about some relevant hardware that is going to need delays of over 1/4 of a microsecond then I am open to putting them at uint16_t. I thought about doing this but the reason I put at uint8_t is because there are also some other CS related timing parameters that people might want to add in the future. But maybe not since I think many of them such as like pcs-pcs delay are not relevant to zephyr API since we expect CS asserted through whole spi_buf_set.

> Another issue as well: how device drivers are going to know which from the delay to the anonymous struct to use? there is no flag anywhere for mitigation. Unless we want to move to a new delay system, which case we will need to make delay attribute obsolete etc...

Actually, technically this is not an issue, but does require a change on this PR, which is on my TODO list in the PR message, maybe I didn't explain it well or should have finished the work before submitting the PR, but I'm just low on bandwidth so wanted the idea reviewed before I did it. 

Basically this information about the chip select pretty much always comes from DT already right now, for GPIO CS. And there is no source of information for native CS.  GPIO cs information comes from the cs-gpios property on the DT node and is initialized by the SPI_DT_SPEC_GET macro. That macro can just be edited to use initialize the struct using the new spi timing properties on the peripheral nodes if they are there and there is no cs-gpios, to initialize the struct to be gpio port as null and use the hw timing members of the union

decsny commented:
I forgot to put DNM since this PR is really only a WIP draft but I opened it as non-draft to get feedback on the idea from relevant reviewers

tbursztyka commented:
> Basically this information about the chip select pretty much always comes from DT already right now, for GPIO CS. And there is no source of information for native CS.

The source of native CS is reg in spi-device.yaml. Not very much intuitive I agree as is also selects the right slot of cs-gpios when this one is present.

decsny commented:
> > Basically this information about the chip select pretty much always comes from DT already right now, for GPIO CS. And there is no source of information for native CS.
> 
> The source of native CS is reg in spi-device.yaml. Not very much intuitive I agree as is also selects the right slot of cs-gpios when this one is present.

I'm not seeing anything there right now, which property are you talking about? To be clear I am saying I will add it on this PR if we agree about this. Right now I just see `spi-cpol`, `spi-cpha`, `spi-hold-cs`,`frame-format`,`spi-max-frequency`, and `duplex` on the spi-device.yaml. Maybe you are thinking about the spi-controller.yaml ?

To be clear, the main point of what I am saying here is that the spi clock timing parameters should be defined on the *device/slave/peripheral* DT node, and **not** the *controller/master* node. Right now we have at least spi-max-frequency on the SPI device nodes, but we also should put these other timing parameters which would come from the datasheets of spi devices to describe the digital AC characteristics of the signals

decsny commented on include/zephyr/drivers/spi.h:
```diff
@@ -152,24 +152,49 @@ extern "C" {
+			/** @brief Delay from PCS assertion to first SCK edge in nanoseconds.
+			 * If set to 0, hardware default will be used.
+			 */
+			uint8_t pcs_to_sck_delay_ns;
```
For NXP, the register programming model is also relative to an equation relating some clock speed and transfer prescalers, and not nanoseconds. But since this is generic API it seems better to use a standard unit which would mean the same no matter the platform, and driver can calculate how to program hardware.

cujomalainey replied:
Then that follows up that we should create some generic helpers if common platforms are going to be doing the same math

decsny replied:
I don't think so, because the math is completely platform specific, that's the point of why I want to specify in ns in the first place

decsny commented:
FYI I plan to flesh out this PR based on discussion and remaining TODO I mentioned, but not a top priority for a few weeks probably

decsny commented:
I think I am going to try to work on this PR again, I think the goal specifically should be to associate timing parameters with the peripherals of the bus, as opposed to the controller, and then have the controller be able to tell what it should do based on what the peripheral needs. That's what is actually important I think. And not worrying so much about native vs gpio control etc.

Commit:
dts: bindings: spi-device: Add parameters for spi peripherals

Add spi parameters for spi peripherals to spi-device.yaml.

These properties and their names are mostly inspired by linux
spi-peripheral-props.yaml.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: Convert to use SPI macro without delay parameters

Convert all drivers and other consumers to use SPI macros without the
delay parameters.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
include: drivers: spi.h: Get params from DT

- Get bit ordering from DT property in SPI_CONFIG_DT.
  More precisely, set LSB first if set in DT.

- Get CS polarity from DT property in SPI_CONFIG_DT.

- Add inter-word delay parameter to spi_config struct. If value is 0, the
  value will be half of the period.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: spi: nxp_lpspi: Use spi_config timing params

Use the timing params from spi_config that are specific to the slave
instead of using the same timing for the controller for all slaves.
Remove these properties from the LPSPI DT binding.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
tests: spi_loopback: Add timing params to RT1050 overlay

Tested case of using these new properties

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

decsny commented:
@tbursztyka the PR is not quite finished but it is almost there I think. And the idea is much more mature now and I already completed a lot of the rework, so would appreciate your re-review

decsny commented:
@henrikbrixandersen would this need to go to Arch WG since it is a breaking change of stable API

henrikbrixandersen commented:
> @henrikbrixandersen would this need to go to Arch WG since it is a breaking change of stable API

Yes, it needs to follow the proces described here: https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-breaking-api-changes

tbursztyka commented:
> @tbursztyka , the process brix linked above looks pretty heavy. I think the only breaking change is the number of arguments to the macro. Would it make sense to keep the macro the same and introduce a new macro that doesnt have the delay parameter, or is it worth making this breaking change for consistency?

I overlooked this aspect, but in any case you will need a new macro and mark the old one obsolete. I don't think you can avoid the arch wg then. (I should be able to join, I'll try to follow the agenda)

decsny commented:
I am not trying to avoid the arch wg but avoid the TSC xD. I think the only breaking change is the macro, so keeping the old one with 3 arguments and obsoleting it and using a new one with two arguments would make it a non breaking API change

decsny commented:
Actually, I think having a bunch of almost duplicate macros that nobody is really intended to use anyways is kind of stupid, especially since downstream users will want to switch to the new macros anyways I would imagine, so I guess I will go through the TSC process once I get this PR passing CI

decsny commented:
@tbursztyka do you still have a change request that isn't addressed?

swift-tk commented:
I thought the spi drivers typically uses gpio cs, i.e. software cs instead of hardware cs? If so, the time to set gpio high or low may already be dozens of nanoseconds?

tbursztyka commented:
> @tbursztyka this will be in arch WG meeting tomorrow, can you attend ?

Yes

decsny commented:
I attempted to implement what Bjarki said in the arch WG meeting call, this is the result:

`west build -b nrf52840dk/nrf52840 -p always zephyr/tests/drivers/spi/spi_loopback/ -DCONFIG_COMPILER_SAVE_TEMPS=y`:

Without delay parameter (new behavior, complex but correct):
`.delay = ((((((0) > (0)) ? (0) : (0))) + (1000) - 1) / (1000))`

Setting delay parameter in the test to some random value (25):
`.delay = 25`

So I guess the type of trick suggested by @bjarki-andreasen  works, and if people are generally agreeing with @carlescufi that the only breaking change is in these macros, and that the spi config is not expected to be made other ways, do I remove the breaking API label ?

decsny commented:
@carlescufi on the push I just did I figured out how to make a warning appear if the delay parameter is provided in any of the macros with just a one line change, thanks to some obscure macro in the toolchain headers

bjarki-andreasen reviewed (changes requested):
Change request is just to move the commit with the spi macro changes, macros look great!

decsny commented on include/zephyr/drivers/spi.h:
```diff
@@ -310,6 +329,22 @@ struct spi_cs_control {
+	MAX(DT_PROP_OR(node_id, spi_cs_setup_delay_ns, 0),	\
+	    DT_PROP_OR(node_id, spi_cs_hold_delay_ns, 0))
+
+#define SPI_CS_CONTROL_INIT_GPIO_DELAY_LEGACY(delay) delay __DEPRECATED_MACRO
+
+#define SPI_CS_CONTROL_INIT_GPIO(node_id, ...)						\
+	.gpio = SPI_CS_GPIOS_DT_SPEC_GET(node_id),					\
+	.delay = COND_CODE_1(IS_EMPTY(__VA_ARGS__),					\
+			(DIV_ROUND_UP(SPI_CS_CONTROL_MAX_DELAY(node_id), 1000)),	\
+			(SPI_CS_CONTROL_INIT_GPIO_DELAY_LEGACY(__VA_ARGS__)))
```
Well, all these changes to the API regarding the CS were originally in the same commit as all the 100+ file changes due to bisectability , should I split it out now?

Commit:
include: drivers: spi.h: Get CS delay parameters from DT

The CS delay parameter did not make a distinction between the setup and
hold time of the CS, and also did not specify very fine control which
can be done usually by a native controller CS. So use the new nanosecond
DT properties to get the delay values and make distinction.

Add deprecation warning if consumer supplies the delay parameter and
make it still work the old way in that case for backward compatibility
following API lifecycle process.

Update driver API version to 1.1.0 due to this change

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

kartben reviewed (changes requested):
Needs to be mentioned in "API Changes" in the release notes, as per https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#deprecated

kartben commented on doc/releases/migration-guide-4.3.rst:
```diff
@@ -73,6 +73,17 @@ Phy
+SPI
+===
+
+* The macros :c:macro:`SPI_CS_CONTROL_INIT` :c:macro:`SPI_CS_CONTROL_INIT_INST`,
```
```suggestion
* The macros :c:macro:`SPI_CS_CONTROL_INIT`, :c:macro:`SPI_CS_CONTROL_INIT_INST`,
```

erwango commented on dts/bindings/spi/spi-device.yaml:
```diff
@@ -58,3 +61,22 @@ properties:
+    type: int
+    description: |
+     Delay in nanoseconds to be introduced by the controller before the CS is de-asserted.
+      Also known as enable lag time.
```
Non blocking
```suggestion
     Also known as enable lag time.
```
Also, maybe (still non blocking):
```suggestion
     Also known as "enable lag time".
```
To be applied above as well.

erwango commented on include/zephyr/drivers/spi.h:
```diff
@@ -242,19 +242,38 @@ extern "C" {
+			 * CS enable lead time, i.e. how long should the CS be asserted
+			 * before the first clock. Specified in nanoseconds.
+			 */
+			uint32_t lead;
```
Why not reusing same terminology than the new bindings ?
s/lead/cs_setup
s/lag/cs_hold

Introducing new terminology makes things confusing IMO, and will complicate searching through the code base.

decsny replied:
There was not old terminology because this was never supported, not sure what you're talking about

erwango replied:
I mean terminology used in the device tree bindings (hold-on, set-up)

tbursztyka replied:
I think  @erwango has a point here, setup_delay/hold_delay are indeed more self-documented than lead/lag 
(plus they directly relate to their respective dts  counterparts)

decsny commented:
@tbursztyka can you re-review and also check if you agree to erwan's suggestion above as youre maintainer/assignee

decsny commented:
@tbursztyka @bjarki-andreasen @erwango @kartben review preferably before next merge conflict

jhedberg commented:
@decsny looks like this caused a CI failure in main: https://github.com/zephyrproject-rtos/zephyr/actions/runs/18160982496/job/51691443328

```
zephyr/drivers/sensor/pni/rm3100/CMakeFiles/drivers__sensor__pni__rm3100.dir/rm3100.c.obj -MF zephyr/drivers/sensor/pni/rm3100/CMakeFiles/drivers__sensor__pni__rm3100.dir/rm3100.c.obj.d -o zephyr/drivers/sensor/pni/rm3100/CMakeFiles/drivers__sensor__pni__rm3100.dir/rm3100.c.obj -c /__w/zephyr/zephyr/drivers/sensor/pni/rm3100/rm3100.c
/__w/zephyr/zephyr/drivers/sensor/pni/rm3100/rm3100.c:262:29: error: too many arguments provided to function-like macro invocation
  262 | DT_INST_FOREACH_STATUS_OKAY(RM3100_DEFINE)
      |                             ^
/__w/zephyr/zephyr/include/zephyr/drivers/spi.h:1364:9: note: macro 'SPI_DT_IODEV_DEFINE' defined here
 1364 | #define SPI_DT_IODEV_DEFINE(name, node_id, operation_)                          \
      |         ^
1 error generated.
ninja: build stopped: subcommand failed.
```

jhedberg commented:
> @decsny looks like this caused a CI failure in main: https://github.com/zephyrproject-rtos/zephyr/actions/runs/18160982496/job/51691443328

My attempt at fixing this: #96859
